File: week01/credit/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int input_length;
long card_number;
int calc_checksum(int digits_left);
int calc_input_length(long number);
bool validate_input_length(int length);
string identify_card_type(long card_number, int input_length);

int main(void)
{
    // get card number
    card_number = get_long("Number: ");

    // calculate length of number 
    input_length = calc_input_length(card_number);

    // check input length
    if (!validate_input_length(input_length))
    {
        printf("INVALID\n");
        return 0;
    }

    // calculate checksum
    int checksum = calc_checksum(1);
    if (!(checksum % 10 == 0)) {
        printf("INVALID\n");
        return 0;
    }

    // identify & print card type
    string card_type = identify_card_type(card_number, input_length);
    printf("%s\n", card_type);
}

int calc_input_length(long number)
{
    if (number < 10)
    {
        return 1;
    }
    return 1 + calc_input_length(number / 10);
}

bool validate_input_length(int length)
{
    switch(length) {
        case 13:
        case 15:
        case 16:
            return true;
            break;
        default:
            return false;
    }
}

int calc_checksum(int current_position)
{
    int mod = powl(10, current_position);
    // get current digit
    int digit = card_number % mod / mod;

    // calculate number to add to checksum
    int add = digit;
    if ((input_length - current_position) % 2 == 1)
    {
        add = add * 2;
    }
    add = (add / 10) + (add % 10);

    // end recursion, when at first position
    if (current_position == input_length)
    {
        return add;
    }

    // calculate next position
    return add + calc_checksum(current_position + 1);
}

string identify_card_type(long card_number, int input_length)
{
    int first_digit = card_number / powl(10, input_length - 1);
    int first_two_digits = card_number / powl(10, input_length - 2);

    switch(input_length) {
        case 13:
            if (first_digit == 4)
            {
                return "VISA";
            }
            break;
        case 15:
            if (first_two_digits == 34 || first_two_digits == 37)
            {
                return "AMEX";
            }
            break;
        case 16:
            if (first_two_digits >= 51 && first_two_digits <= 55)
            {
                return "MASTERCARD";
            }
            if (first_digit == 4)
            {
                return "VISA";
            }
            break;
    }

    return "INVALID";
}

File: week06/credit/credit.py
from cs50 import get_int


VALID_CARDNUMBER_LENGTHS = [13, 15, 16]


def main():
    cardnumber = get_int("Number: ")
    cardnumber_length = calc_length_of_cardnumber(cardnumber)

    if not cardnumber_length in VALID_CARDNUMBER_LENGTHS:
        print("INVALID")
        return

    checksum = calc_checksum_of_cardnumber(cardnumber, cardnumber_length)

    if not checksum % 10 == 0:
        print("INVALID")
        return

    card_type = identify_card_type(cardnumber, cardnumber_length)
    print(card_type)


def calc_length_of_cardnumber(number):
    return len(str(number))


def calc_checksum_of_cardnumber(number, length):
    checksum = 0
    number = str(number)
    for i in range(0, len(number)):
        is_even = (length - i) % 2 == 0
        digit = int(number[i])
        checksum += add_to_checksum(number[i], is_even, length)

    return checksum


def add_to_checksum(digit, is_even, length):
    digit = int(digit)
    if is_even:
        digit *= 2
    if digit > 9:
        digit = sum_digits(digit)
    return digit


def sum_digits(number):
    sum = 0
    while number > 0:
        sum += number % 10
        number //= 10

    return sum


def identify_card_type(number, length):
    if is_VISA(number, length):
        return "VISA"
    elif is_AMEX(number, length):
        return "AMEX"
    elif is_MASTERCARD(number, length):
        return "MASTERCARD"
    else:
        return "INVALID"


def is_VISA(number, length):
    first_digit = int(str(number)[0])
    if first_digit == 4 and length in [13, 16]:
        return True
    return False


def is_AMEX(number, length):
    first_two_digits = int(str(number)[0:2])
    if first_two_digits in [34, 37] and length == 15:
        return True
    return False


def is_MASTERCARD(number, length):
    first_two_digits = int(str(number)[0:2])
    if first_two_digits in [51, 52, 53, 54, 55] and length == 16:
        return True
    return False


main()